docs/architecture.md

```diff
@@ -0,0 +1,331 @@
+```rust
+impl fuel_poa_consensus::ports::BlockProducer for Service<BlockProducer> {
+    // insert impl here
+}
+```
+
+### Block Importer
+
+#### Ports: fuel_block_importer::ports
+
+```rust
+trait Relayer {
+    // wait until the relayer is synced with ethereum
+    async fn await_synced() -> Result<()>;
+    // wait until a specific da_height is reached
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+}
+
+trait Executor {
+    async fn validate_and_store_block(block: &FuelBlock) -> Result<()>;
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+```rust
+trait PeerToPeer {
+    async fn fetch_best_peer_height() -> Result<(PeerId, BlockHeight)>;
+    async fn fetch_block_headers(query: Range<BlockHeight>) -> Result<Vec<(PeerId, SealedFuelBlockHeader)>>;
+    async fn fetch_block_data(id: block_id) -> Result<Vec<(PeerId, FuelBlockBody)>>;
+    async fn report_invalid_height(peer: PeerId) -> Result<()>;
+    async fn report_invalid_block(peer: PeerId) -> Result<()>;
+}
+
+trait BlockImporter {
+    // commit a sealed block to the chain
+    async fn commit(block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Consensus {
+    // check with the consensus layer whether a block header passes consensus rules
+    async fn validate_sealed_block_header(block: SealedFuelBlockHeader) -> Result<()>;
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    async fn broadcast_transaction(transaction: Transaction) -> Result<()>;
+    fn report_invalid_tx(peer_id: PeerId) -> Result<()>;
+    // handler for gossipped transactions
+    fn on_new_tx(tx: Transaction, peer: PeerId);
+}
+```
+
```

Comment:
Does fuel-txpool need anything else specifying how blocks are produced? I was thinking the other day it might be nice to have some way for block production to be externally delegated to something over a public interface if made accessible. This greatly helps with a PBS or similar structure, however would then require changes in the public interface.

I think it could be very valuable to allow TxPool to delegate a `MakeBlock` function so that custom block building can easily integrate inside the txpool. Although we could also undertake this with a similar internal change of just having some `BlockProduction` Interface, and allowing people to swap in their own custom production methods via that.

I think it's worth weighing the 2 options, as a public interface allows more complex block building (especially if someone has tx's their txpool doesn't know about yet with a txpool aggregator across peergroups) and it also helps with something like PBS, but an inner strategy function avoids needing to have a public interface and is more simplistic.